### ISSUE TITLE
New workflow, precomputed static cell set visualization

### ISSUE THREAD

**Issue by jimmymathews** [nadeemlab/SPT#119]
For this issue,

- Add a new workflow that computes UMAP visualization of the entire cell set for a given study, along the phenotypes that are part of the study.
- Add a new table ("out of band", i.e. not part of the main data model) with a base64 encoding of SVG or similar vector graphics representations of: one UMAP for each channel with an overlay of the intensity channel (say, in green to gray).
- Then add an API endpoint that retrieves all the UMAP base64 strings by study.

Note that, to enable the above, first the data import workflow needs to be updated to retrieve the intensity values for each channel and cell (See #125, which blocks this).

(This work is arguably blocked by #83, since #83 should provide a clean template for new workflows.)

**Comment by danjafish** [comment]
@jimmymathews do you expect me to start with the issue #125 or work on this one assuming some dummy values?

**Comment by danjafish** [comment]
@jimmymathews That's an example of UMAP clustering for sample dataset. Here are 7 samples and separate cluster for each sample. But I assume that's not always a case. Can you clarify cluster coloring? What should we use as clusters? 
![image](https://user-images.githubusercontent.com/34601963/231773158-3c4eb91f-54e3-4f99-8ef0-4bab81e12aa5.png)

**Comment by danjafish** [comment]
@jimmymathews I've created a new branch [issue119](https://github.com/nadeemlab/SPT/tree/issue119). The changes are quite significant, so I suggest reviewing them before I submit a PR. Here are a few notes:

The changes have not been tested on large datasets yet.
I've added an API endpoint based on your previous work as an example, but it hasn't been tested. This is another issue we should address.
There are several assumptions made about the workflow structure, so please feel free to provide feedback if you think something needs to be changed.
We should consider adding tests, as there aren't any at the moment.

**Comment by jimmymathews** [comment]
@danjafish : I created `issue119review` branched off of `issue119`. I refactored a bit and added a test that creates an HTML page out of all the images, and compares it with such a page committed to version control.

Please have a look and merge it into `issue119` if you're satisfied, that is:

```sh
git checkout issue119
git merge issue119review
```

Then when you're ready, you can make a PR for `issue119`.

**Comment by jimmymathews** [comment]
(Just realized I didn't look at the API endpoint yet.)

**Comment by jimmymathews** [comment]
A major thing I changed is to drop the use of the "quantitative feature value" system for these. The intended use for that was sample-level features, whereas these UMAP plots are images (not quantitative values) associated with a whole study (not sample-level).

**Comment by jimmymathews** [comment]
I also moved the creation of the new table to the `spatialprofilingtoolbox/db/performance...sql` portion of the SPT-specific data model.

**Comment by jimmymathews** [comment]
I have taken the liberty of merging `issue119review` into `issue119` (I hesitated to do so earlier because this makes it slightly harder for you to compare your last commit with the updated version).

I reviewed the new API endpoint and added a test for it. It's not shown in the code for the test, but I checked that at least one of the test data artifact base64 strings in the applicable case is a valid SVG, and the figure looks right.

The test suite passes for me.

**Comment by jimmymathews** [comment]
I ran this on the big datasets and it crashed (on 2.5m cells and 26 channels, exceed my 32GB memory) and "worked" in another case: 500k cells and 33 channels, takes 25 minutes and peak RAM usage around 18GB.

I also realized that it's pretty crazy to save vector graphics formats for millions of cells. The SVG file is in the hundreds of megabytes. And there is no real benefit, because most cells will not be seen. The ideal thing wouldbe: A vector format which has the same on-screen appearance as the huge SVG, but the easier way is normal rasterization at fairly high resolution.

So downsampling and rasterization seem to be the right thing to do. I think 50k or so would be more than adequate for visualization purposes.

**Comment by jimmymathews** [comment]
The last feature (hopefully the last) I'd like to implement as part of this issue is to limit the SELECT query, that pulls the intensity data, to a random subset of cells. Currently the random subset takes places after the database pull, which is pretty bad for bandwidth and unnecessary strain on the database.

**Comment by jimmymathews** [comment]
I'm having trouble getting the umap creation to be truly deterministic. There seem to be some machine-dependent steps. I will just remove the final check that the exact PNG is reproduced.

**Comment by jimmymathews** [comment]
A few updates based on runs with big datasets:
- The marker size needed to be decreased.
- The color range needed to be expanded.
- Made "overwrite old plots" the default behavior for re-running of the workflow.
- Added a convenience script for creating a plots HTML page preview.

**Comment by jimmymathews** [comment]
Re-opening briefly to deal with the following issue: The `visualization-plots` endpoint comes back with about 100MB payload. The high-res images (2400 x 2400) are good but these should be provided one-by-one on request. The API  server should send back much smaller resolution images, and a separate endpoint can be added for the full-res.

**Comment by jimmymathews** [comment]
The large size is partly due to the base64 encoding, but this encoding is important for being able to make a single JSON payload rather than several dozen individual PNG file responses.
This has been implemented in branch `issue119`.
